src/network/src/esp/src/NetworkBroadcast.cpp

```diff
@@ -0,0 +1,103 @@
+#include "NetworkBroadcast.h"
+#include "NetworkConfig.h"
+#include "SocketFactory.h"
+#include "Task.h"
+#include "lwip/sockets.h"
+
+static void task(void* context) {
+    while (true) {
+        static_cast<NetworkBroadcast*>(context)->receiveDatagrams();
+        Task::delay(20);
+    }
+}
+
+NetworkBroadcast::NetworkBroadcast(ILogger& logger) :
+    m_logger(logger),
+    m_socket(NO_SOCKET),
+    m_started(false),
+    m_receivingTask("udp_receive", tskIDLE_PRIORITY, task, this) {
+    CircularBuff_init(&m_circularBuffer, m_data.data(), m_data.size());
+    m_receivingTaskHandle = nullptr;
+    m_receivingTask.start();
+}
+
+NetworkBroadcast::~NetworkBroadcast() { this->stop(); }
+
+bool NetworkBroadcast::start() {
+    if (!m_started) {
+        m_socket = SocketFactory::createUDPBroadcast(NetworkConfig::getBroadcastInputPort());
+        if (m_socket < 0) {
+            m_logger.log(LogLevel::Error, "Failed to obtain broadcast socket!");
+            return false;
+        }
+        m_started = true;
+    }
+    return true;
+}
+
+bool NetworkBroadcast::send(const uint8_t* data, uint16_t length) {
+    if (data == nullptr || m_socket < 0) {
+        return false;
+    }
+
+    // Destination broadcast
+    sockaddr_in broadcast{};
+    broadcast.sin_addr.s_addr = htonl(INADDR_BROADCAST);
+    broadcast.sin_family = AF_INET;
+    broadcast.sin_port = htons(NetworkConfig::getBroadcastOutputPort());
+    broadcast.sin_len = sizeof(broadcast);
+
+    // Use lwip_sendto with udp since it is a connection-less protocol.
+    return lwip_sendto(m_socket, data, length, 0, (sockaddr*)&broadcast, sizeof(broadcast)) ==
+           length;
+}
+
+bool NetworkBroadcast::receive(uint8_t* data, uint16_t length) {
+    if (data == nullptr || m_socket < 0) {
+        return false;
+    }
+
+    m_receivingTaskHandle = xTaskGetCurrentTaskHandle();
+    while (CircularBuff_getLength(&m_circularBuffer) < length) {
+        ulTaskNotifyTake(pdTRUE, 500);
+    }
+    m_receivingTaskHandle = nullptr;
+
+    return CircularBuff_get(&m_circularBuffer, data, length) == length;
+}
+
+bool NetworkBroadcast::stop() {
+    if (m_socket != -1) {
+        lwip_close(m_socket);
+    }
+    m_socket = NO_SOCKET;
+    m_started = false;
+    return true;
+}
+
+void NetworkBroadcast::receiveDatagrams() {
+    if (m_started) {
+
+        // Receiving from a UDP socket de-queues the whole datagram from the receiving queue,
+        // regardless of the size read. Here, we receive up to a maximum of the size of m_datagram.
+        // The MSG_TRUNC flag will make the lwip_recvfrom return the size of the first datagram in
+        // the queue, and MSG_PEEK insures we do not pop the datagram yet from the queue.
+        // TLDR, the following line gets the size of the first datagram in the socket queue.
+        ssize_t sizeToReceive =
+            lwip_recv(m_socket, m_datagram.data(), m_datagram.size(), MSG_TRUNC | MSG_PEEK);
```

Comment:
Why do we need to know the size first? Why not just read a datagram of maximum size of m_datagram.size()? I would have though it would be for error handling if the datagram is bigger thand the size of the buffer, but that does not seem to be the case.

Also what would happen if the datagram is bigger than the buffer? Would we lose some data?

Reply:
This is weirdly explained in the doc about the socket flags, but MSG_TRUNC needs to be used alongside MSG_PEEK in order to return on each datagram received. So basically, the peeking call returns the size of a datagram on reception and then we read exactly this size. If we were to read less or more bytes than the size of the datagram, we would be losing data. 

It's a bit finicky, but safe regarding data loss.

Reply:
There is a define `Message_size` in nanopb, as of now, it's 327 bytes. Though the whole message can be 331 bytes since it has a varint header of 4 bytes. I think 400 to 512 bytes would be safe.